test/Microsoft.AspNetCore.Blazor.Build.Test/TestFiles/DesignTimeCodeGenerationTest/Component_WithRef_CSharpReservedWord/TestComponent.codegen.cs

```diff
@@ -0,0 +1,47 @@
+// <auto-generated/>
+#pragma warning disable 1591
+namespace Test
+{
+    #line hidden
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    public class TestComponent : Microsoft.AspNetCore.Blazor.Components.BlazorComponent
+    {
+        #pragma warning disable 219
+        private void __RazorDirectiveTokenHelpers__() {
+        ((System.Action)(() => {
+global::System.Object __typeHelper = "*, TestAssembly";
+        }
+        ))();
+        }
+        #pragma warning restore 219
+        #pragma warning disable 0414
+        private static System.Object __o = null;
+        #pragma warning restore 0414
+        #pragma warning disable 1998
+        protected override void BuildRenderTree(Microsoft.AspNetCore.Blazor.RenderTree.RenderTreeBuilder builder)
+        {
+            base.BuildRenderTree(builder);
+#line 2 "x:\dir\subdir\Test\TestComponent.cshtml"
+                                         @class = default(Test.MyComponent);
```

Comment:
> The ref tag helper does show correct warning squigglies if you reference a nonexistent variable (e.g., <div ref="somethingThatIsNotDefined" />). But it does not show squigglies if you reference something of the wrong type (e.g., <div ref="someIntField" />). 

I see now what you mean about the errors doing the wrong thing. If you want to address that problem, I'd suggest generating something like.

```
BindMethods.AssignComponent<Test.MyComponent>(
#line .....
              @class
#line default
#line hidden
);
@class = default;
```

Where you've defined:
```
public void BindMethods.AssignComponent<T>(T value)
```

This would help in a few ways:
1. It provides a context where the type of the user-provided expression has to be known
2. It makes sure the value gets assigned at some point, preventing a warning.

How this could break down is if the expression provided is not an LValue. We already have this problem with bind, and I'm not sure there's a 100% perfect solution without more love from C#.

Reply:
I did try something along these lines yesterday but it led to the allowed conversions being the wrong way round:

    SomeMethod(MyComponent instance)

... allows you to pass anything that implicitly converts *to* a `MyComponent`, but I need an expression that allows you to pass anything that implicitly *from* `MyComponent`.

I also experimented with `SomeMethod(out MyComponent instance)`, but that doesn't work because `out` (like `ref`) only matches exact types, not things that are implicitly convertible.

AFAICT the only construction that has the correct behavior is an assignment (`yourIdentifier = (MyComponent)anything;`), which is what comes with the limitation that it can only catch one of the two types of errors.

If I'm missing something obvious please let me know!

Reply:
That makes sense. There are limitations here that we can't really avoid.

BTW WPF solves this problem by not letting you write the field declaration - which means that it's always a field and always private.